cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/RemoteProgramRunDispatcher.java

```diff
@@ -0,0 +1,131 @@
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteProgramRunDispatcher.class);
+
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(
+      new GsonBuilder())
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .registerTypeAdapter(ApplicationClass.class, new ApplicationClassCodec())
+    .registerTypeAdapter(Requirements.class, new RequirementsCodec())
+    .registerTypeAdapter(RunId.class, new RunIds.RunIdCodec())
+    .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+    .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
+    .create();
+  private final Store store;
+  private final ProgramRunnerFactory programRunnerFactory;
+  private final RemoteTaskExecutor remoteTaskExecutor;
+  private final ProgramRunnerFactory remoteProgramRunnerFactory;
+  private final RemoteExecutionTwillRunnerService remoteTwillRunnerService;
+
+  @Inject
+  public RemoteProgramRunDispatcher(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
+                                    RemoteClientFactory remoteClientFactory, Store store,
+                                    ProgramRunnerFactory programRunnerFactory,
+                                    @RemoteExecution ProgramRunnerFactory remoteProgramRunnerFactory,
+                                    @RemoteExecution RemoteExecutionTwillRunnerService remoteTwillRunnerService) {
+    this.store = store;
+    this.programRunnerFactory = programRunnerFactory;
+    this.remoteProgramRunnerFactory = remoteProgramRunnerFactory;
+    this.remoteTwillRunnerService = remoteTwillRunnerService;
+    this.remoteTaskExecutor = new RemoteTaskExecutor(cConf, metricsCollectionService,
+                                                     remoteClientFactory, RemoteTaskExecutor.Type.SYSTEM_WORKER);
+  }
+
+  @Override
+  public ProgramController dispatchProgram(ProgramRunDispatcherInfo programRunDispatcherInfo) throws Exception {
+    RunId runId = programRunDispatcherInfo.getRunId();
+    LOG.debug("Dispatching Program Run operation for Run ID: {}", runId.getId());
+    RunnableTaskRequest request = RunnableTaskRequest.getBuilder(ProgramRunDispatcherTask.class.getName())
+      .withParam(GSON.toJson(programRunDispatcherInfo)).build();
+    byte[] result = remoteTaskExecutor.runTask(request);
+    RemoteProgramRunDispatcherResponse response =
+      GSON.fromJson(new String(result, StandardCharsets.UTF_8), RemoteProgramRunDispatcherResponse.class);
+    if (!response.isProgramRunSuccessful()) {
+      // The response should always be true. Otherwise, callee will throw an exception.
+      throw new Exception("Unexpected exception");
+    }
+    ProgramId programId = programRunDispatcherInfo.getProgramDescriptor().getProgramId();
+    ProgramRunId programRunId = programId.run(runId);
+    ProgramRunner runner =
+      (ProgramRunners.getClusterMode(programRunDispatcherInfo.getProgramOptions()) == ClusterMode.ON_PREMISE
```

Comment:
parentheses here are not needed

Reply:
This piece is identical to the way it was done previously. Haven't changed it. Also, it is needed since the code between (ProgramRunners.getClusterMode(programRunDispatcherInfo.getProgramOptions()) == ClusterMode.ON_PREMISE
        ? programRunnerFactory
        : Optional.ofNullable(remoteProgramRunnerFactory).orElseThrow(UnsupportedOperationException::new)) is returning ProgramRunnerFactory, on which the .create method is used.

Reply:
oh... that's hard to read. Can you break this up into 2 statements? Just because it was done a certain way before doesn't mean it can't be tweaked for readability.